examples/coco_caption/captioner.py

```diff
@@ -0,0 +1,402 @@
+#!/usr/bin/env python
+
+from collections import OrderedDict
+import h5py
+import math
+import matplotlib.pyplot as plt
+import numpy as np
+import os
+import random
+import sys
+
+sys.path.append('./python/')
+import caffe
+
+class Captioner():
+  def __init__(self, weights_path, image_net_proto, lstm_net_proto,
+               vocab_path, device_id=-1):
+    if device_id >= 0:
+      caffe.set_mode_gpu()
+      caffe.set_device(device_id)
+    else:
+      caffe.set_mode_cpu()
+    # Setup image processing net.
+    phase = caffe.TEST
+    self.image_net = caffe.Net(image_net_proto, weights_path, phase)
+    image_data_shape = self.image_net.blobs['data'].data.shape
+    self.transformer = caffe.io.Transformer({'data': image_data_shape})
+    channel_mean = np.zeros(image_data_shape[1:])
+    channel_mean_values = [104, 117, 123]
+    assert channel_mean.shape[0] == len(channel_mean_values)
+    for channel_index, mean_val in enumerate(channel_mean_values):
+      channel_mean[channel_index, ...] = mean_val
+    self.transformer.set_mean('data', channel_mean)
+    self.transformer.set_channel_swap('data', (2, 1, 0))
+    self.transformer.set_transpose('data', (2, 0, 1))
+    # Setup sentence prediction net.
+    self.lstm_net = caffe.Net(lstm_net_proto, weights_path, phase)
+    self.vocab = ['<EOS>']
+    with open(vocab_path, 'r') as vocab_file:
+      self.vocab += [word.strip() for word in vocab_file.readlines()]
+    net_vocab_size = self.lstm_net.blobs['predict'].data.shape[2]
+    if len(self.vocab) != net_vocab_size:
+      raise Exception('Invalid vocab file: contains %d words; '
+          'net expects vocab with %d words' % (len(self.vocab), net_vocab_size))
+
+  def set_image_batch_size(self, batch_size):
+    self.image_net.blobs['data'].reshape(batch_size,
+        *self.image_net.blobs['data'].data.shape[1:])
+
+  def caption_batch_size(self):
+    return self.lstm_net.blobs['cont_sentence'].data.shape[1]
+
+  def set_caption_batch_size(self, batch_size):
+    self.lstm_net.blobs['cont_sentence'].reshape(1, batch_size)
+    self.lstm_net.blobs['input_sentence'].reshape(1, batch_size)
+    self.lstm_net.blobs['image_features'].reshape(batch_size,
+        *self.lstm_net.blobs['image_features'].data.shape[1:])
+    self.lstm_net.reshape()
+
+  def preprocess_image(self, image, verbose=False):
+    if type(image) in (str, unicode):
+      image = plt.imread(image)
```

Comment:
I just tried to use it today to predict captions on some PNG images, and it doesn't work, because for png images, `plt.imread` returns (0, 1) np.float32 rather than uint8.